spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/CloudVisionTemplate.java

```diff
@@ -0,0 +1,107 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.vision;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.BatchAnnotateImagesRequest;
+import com.google.cloud.vision.v1.BatchAnnotateImagesResponse;
+import com.google.cloud.vision.v1.Feature;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.protobuf.ByteString;
+
+import org.springframework.core.io.Resource;
+
+/**
+ * Spring Template offering convenience methods for interacting with the Cloud Vision
+ * APIs.
+ *
+ * @author Daniel Zou
+ *
+ * @since 1.1
+ */
+public class CloudVisionTemplate {
+
+	private final ImageAnnotatorClient imageAnnotatorClient;
+
+	public CloudVisionTemplate(ImageAnnotatorClient imageAnnotatorClient) {
+		this.imageAnnotatorClient = imageAnnotatorClient;
+	}
+
+	/**
+	 * Analyze an image and extract the feature of the image specified by {@code featureType}.
+	 *
+	 * <p>
+	 * A feature describes the kind of Cloud Vision analysis one wishes to perform on an
+	 * image, such as text detection, image labelling, facial detection, etc. A full list of
+	 * feature types can be found in {@link Feature.Type}.
+	 *
+	 * @param imageResource the image one wishes to analyze. The Cloud Vision APIs support
+	 *     image formats described here: https://cloud.google.com/vision/docs/supported-files
+	 * @param featureType the type of image analysis to perform on the image.
+	 * @return the results of image analysis
+	 */
+	public AnnotateImageResponse analyzeImage(Resource imageResource, Feature.Type featureType)
+			throws IOException {
+
+		return analyzeImage(imageResource, Collections.singleton(featureType));
+	}
+
+	/**
+	 * Analyze an image and extract the features of the image specified by
+	 * {@code featureTypes}.
+	 *
+	 * <p>
+	 * A feature describes the kind of Cloud Vision analysis one wishes to perform on an
+	 * image, such as text detection, image labelling, facial detection, etc. A full list of
+	 * feature types can be found in {@link Feature.Type}.
+	 *
+	 * @param imageResource the image one wishes to analyze. The Cloud Vision APIs support
+	 *     image formats described here: https://cloud.google.com/vision/docs/supported-files
+	 * @param featureTypes the types of image analysis to perform on the image.
+	 * @return the results of image analyses
+	 */
+	public AnnotateImageResponse analyzeImage(Resource imageResource, Set<Feature.Type> featureTypes)
```

Comment:
Why can't we have a single method with vararg?

Reply:
Instead of `Set`, does it make sense to use varargs?
```java
public AnnotateImageResponse analyzeImage(Resource imageResource, Feature.Type ... featureTypes)
```

Reply:
Yes I am open to this. Right now this is just a sketch of the template API to be used as a point for discussion as we figure out what the template will look like. I will raise this point in the next discussion we have.